From d21741253c114727d281372e0b8694d59bc172f4 Mon Sep 17 00:00:00 2001
From: Moritz Bunkus <moritz@bunkus.org>
Date: Sat, 7 Jan 2017 16:33:57 +0100
Subject: [PATCH] Rakefile: enable parallel builds with rake 10.1.0 or newer

rake 10.1.0, which is part of Ruby v2.1.x, can do parallel builds, too,
and supports the same option -j that drake does. And it even defaults to
"number of cores + 1".

However, the tasks that can be executed in parallel have to be set up
with "multitask" instead of "task". There is an option to enable
parallel processing for all tasks defined as "tasks", though, and the
commit enables this option.
---
 Rakefile | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/Rakefile b/Rakefile
index 8cd23e4..b8f1f71 100644
--- a/Rakefile
+++ b/Rakefile
@@ -18,11 +18,18 @@ if FileUtils.pwd != File.dirname(__FILE__)
 end
 
 # Set number of threads to use if it is unset and we're running with
-# drake
+# drake.
 if Rake.application.options.respond_to?(:threads) && [nil, 0, 1].include?(Rake.application.options.threads) && !ENV['DRAKETHREADS'].nil?
   Rake.application.options.threads = ENV['DRAKETHREADS'].to_i
 end
 
+# For newer rake turn on parallel processing, too. Newer rake versions
+# use an OpenStruct, though, so testing with responds_to? won't work.
+version = Rake::VERSION.gsub(%r{[^0-9\.]+}, "").split(%r{\.}).map(&:to_i)
+if (version[0] > 10) || ((version[0] == 10) && (version[1] >= 1))
+  Rake.application.options.always_multitask = true
+end
+
 require "pp"
 
 require_relative "rake.d/extensions"
